Add minified Json printing functionality

chezRong opened the pull request:
See https://github.com/google/protobuf/issues/1340
Add ` public Printer minifyingOutput(){
      return new Printer(registry, includingDefaultValueFields, preservingProtoFieldNames, true);
    }` 
method to be used for json formatting, when calling `JsonFormat.printer(). minifyingOutput().print(message)`
the result json output will be minified.

grpc-jenkins commented:
Thanks for your pull request. The automated tests will run as soon as one of the admins verifies this change is ok to test.

xfxyjwf commented on java/util/src/main/java/com/google/protobuf/util/JsonFormat.java:
```diff
@@ -151,7 +158,17 @@ public Printer includingDefaultValueFields() {
+     * The new Printer clones all other configurations from the
+     * current {@link Printer}.
+     */
+    public Printer setPrinterTypeUgly(){
```
I think we need a better name for this option. It should be consistent with the other options.

chezRong replied:
Are you referring to the setPrinterTypeUgly function? Would minifyingJSONOutput be a more suitable name? I guess I should also rename the boolean minified option to match this name change.

xfxyjwf replied:
We care more about the public method name than the internal field name. I'll consult other team members to see what we want this option name to be.

chezRong commented:
I signed it!

dwo commented on .gitignore:
```diff
@@ -80,6 +80,10 @@ src/**/*.trs
+#intellij
+.idea/
+*.iml
+
```
You probably don't want these editor-specific ignores in the pull request.

dwo commented on java/util/src/main/java/com/google/protobuf/util/JsonFormat.java:
```diff
@@ -432,6 +491,8 @@ private void write(final CharSequence data) throws IOException {
     // We use Gson to help handle string escapes.
     private final Gson gson;
+    private final CharSequence nullOrSpace;
+    private final CharSequence nullOrNewLine;
```
maybe `emptyOrSpace` and `emptyOrNewLine` would be a more accurate name for these?

dwo commented:
I signed it!

chezRong commented:
what else needs to done to the CLA? The author has authorized consent

xfxyjwf commented:
@chezRong I discussed with other team members about the name of the public method and we believed "omittingInsignificantWhitespace" is the best name for it. Could you update your pull request accordingly?

chezRong commented:
@xfxyjwf Hi, I have squashed down the commits and updated the code against the final confirmed naming for the method. Thanks:D

xfxyjwf commented on java/util/src/main/java/com/google/protobuf/util/JsonFormat.java:
```diff
@@ -96,42 +96,46 @@
+
   /**
-   * Creates a {@link Printer} with default configurations.
+   * Creates a {@link Printer} with default configurations. Default printing option is Pretty
```
This comment is incomplete. The default configuration is a combination of different options. I suggest we just omit it here and move the detailed explanation to the comment of the omittingInsignificantWhitespace() method.

xfxyjwf commented on java/util/src/main/java/com/google/protobuf/util/JsonFormat.java:
```diff
@@ -96,42 +96,46 @@
   }
-  
+
+
```
Redundant new lines.

xfxyjwf commented on java/util/src/main/java/com/google/protobuf/util/JsonFormat.java:
```diff
@@ -151,12 +155,22 @@ public Printer includingDefaultValueFields() {
+
+
+    /**
+     * Creates a new {@link Printer} that has a different textGenerator, The json output will be omittingInsignificantWhitespace.
```
=> Create a new Printer that will omit all insignificant whitespace in the JSON output. This new Printer clones all other configurations from the current Printer. Insignificant whitespace is defined by the JSON spec as whitespace that appear between JSON structural elements:
<pre>

      ws = *(
              %x20 /              ; Space
              %x09 /              ; Horizontal tab
              %x0A /              ; Line feed or New line
              %x0D )              ; Carriage return
</pre>
See https://tools.ietf.org/html/rfc7159

xfxyjwf commented on java/util/src/main/java/com/google/protobuf/util/JsonFormat.java:
```diff
@@ -360,14 +375,55 @@ private void addMessage(Descriptor message) {
   }
 
   /**
+   * An interface for json formatting that can be used in
```
=> A TextGenerator adds indentation when writing formatted text.

xfxyjwf commented on java/util/src/main/java/com/google/protobuf/util/JsonFormat.java:
```diff
@@ -360,14 +375,55 @@ private void addMessage(Descriptor message) {
+  /**
+   * Format the json without indentation
+   */
+  private static final class UglyTextGenerator implements TextGenerator{
```
UglyTextGenerato => CompatTextGenerator

xfxyjwf commented on java/util/src/main/java/com/google/protobuf/util/JsonFormat.java:
```diff
@@ -441,12 +499,21 @@ private void write(final CharSequence data) throws IOException {
+        this.blankOrSpace = "";
+        this.blankOrNewLine = "";
+      } else {
+          this.generator = new PrettyTextGenerator(jsonOutput);
```
Incorrect indentation.

xfxyjwf commented on java/util/src/main/java/com/google/protobuf/util/JsonFormat.java:
```diff
@@ -581,20 +648,20 @@ private void printAny(MessageOrBuilder message) throws IOException {
+        generator.print("\"value\":" + blankOrSpace);
         printer.print(this, contentMessage);
-        generator.print("\n");
+        generator.print("" + blankOrNewLine);
```
Just "blankOrNewLine". You don't need to append it to an empty string.

xfxyjwf commented on java/util/src/test/java/com/google/protobuf/util/JsonFormatTest.java:
```diff
@@ -139,6 +139,10 @@ private void assertRoundTripEquals(Message message, TypeRegistry registry) throw
     return JsonFormat.printer().print(message);
   }
+  //test for enum UGLY, PRETTY
+  private String toUglyJsonString(Message message) throws IOException{
```
toCompactJsonString

xfxyjwf commented on java/util/src/test/java/com/google/protobuf/util/JsonFormatTest.java:
```diff
@@ -193,6 +197,50 @@ public void testAllFields() throws Exception {
             + "  \"repeatedNestedEnum\": [\"BAR\", \"BAZ\"]\n"
             + "}",
         toJsonString(message));
+    assertEquals(
```
Please add one dedicated test case (one test method) for the new option (instead of spreading it everywhere).

chezRong commented:
@xfxyjwf squashed the commits and changed relevant names to `compactTextGenerator`, updated test file with `testOmittingInsignificantWhiteSpace`. Let me know if there's anything more that needs changing, thanks

xfxyjwf commented on java/util/src/main/java/com/google/protobuf/util/JsonFormat.java:
```diff
@@ -151,9 +153,26 @@ public Printer includingDefaultValueFields() {
-    
+
+    /**
+     * Create a new  {@link Printer}  that will omit all insignificant whitespace in the JSON output. This new Printer clones all other configurations from the current Printer.
```
This line is too long. The column limit is 100 according to [Google Java Style Guide](https://google.github.io/styleguide/javaguide.html#s4.4-column-limit).

xfxyjwf commented on java/util/src/main/java/com/google/protobuf/util/JsonFormat.java:
```diff
@@ -151,9 +153,26 @@ public Printer includingDefaultValueFields() {
+
+    /**
+     * Create a new  {@link Printer}  that will omit all insignificant whitespace in the JSON output. This new Printer clones all other configurations from the current Printer.
+     * The new Printer clones all other configurations from the Insignificant whitespace is defined by the JSON spec as whitespace that appear between JSON structural elements:
```
"The new Printer clones all other configurations from the" is duplicated from the above line.

xfxyjwf commented on java/util/src/test/java/com/google/protobuf/util/JsonFormatTest.java:
```diff
@@ -1171,4 +1174,59 @@ public void testPreservingProtoFieldNames() throws Exception {
+    setAllFields(builder);
+    TestAllTypes message2 = builder.build();
+    assertEquals(
+            "{"
```
Indent this by 4 spaces.

xfxyjwf commented:
Thanks. Looks good. Just a few more style comments.
